spec/models/field_spec.rb

```diff
@@ -0,0 +1,356 @@
+      it "raises an ActiveRecord::StatementInvalid error" do
+        expect{Field.order_by(:invalid_attribute).to_a}.to raise_error(ActiveRecord::StatementInvalid)
+      end
+    end
+
+    context "when provided direction is invalid" do
+      it "raises an ArgumentError error" do
+        expect{Field.order_by(:name, :up)}.to raise_error(ArgumentError)
+      end
+    end
+  end
+
+  describe ".most_rated" do
+    let!(:field_1){create :field}
+    let!(:field_2){create :field}
+    let!(:rating_1){create :rating, field: field_1, rating: 1.0}
+    let!(:rating_2){create :rating, field: field_1, rating: 2.0}
+    let!(:rating_3){create :rating, field: field_2, rating: 4.0}
+    let!(:rating_4){create :rating, field: field_2, rating: 5.0}
+
+    context "when true provided" do
+      it "orders by average rating in descending order" do
+        expect(Field.most_rated true).to eq([field_2, field_1])
+      end
+    end
+
+    context "when falsy or no argument provided" do
+      it "returns in default order" do
+        expect(Field.most_rated false).to eq([field_1, field_2])
+        expect(Field.most_rated).to eq([field_1, field_2])
+      end
+    end
+  end
+
+  describe ".field_type" do
+    let!(:field_type_1){create :field_type}
+    let!(:field_type_2){create :field_type}
+    let!(:field_1){create :field, field_type: field_type_1}
+    let!(:field_2){create :field, field_type: field_type_1}
+    let!(:field_3){create :field, field_type: field_type_2}
+
+    context "when field_type_id provided" do
+      it "returns fields in this field type" do
+        expect(Field.field_type field_type_1.id).to include(field_1, field_2)
+        expect(Field.field_type field_type_2.id).to include(field_3)
+      end
+    end
+
+    context "when 'all' or no argument provided" do
+      it "returns all fields" do
+        expect(Field.field_type "all").to include(field_1, field_2, field_3)
+        expect(Field.field_type).to include(field_1, field_2, field_3)
+      end
+    end
+  end
+
+  describe ".favourite_by_current_user" do
+    let!(:user){create :user}
+    let!(:field_1){create :field}
+    let!(:field_2){create :field}
+    let!(:field_3){create :field}
+
+    before do
+      user.add_favourite_field [field_1, field_2]
+    end
+
+    context "when current user favourite fields ids provied" do
+      it "returns fields current user added to favourite list" do
+        expect(Field.favourite_by_current_user user.favourite_field_ids)
+        .to include(field_1, field_2)
+        expect(Field.favourite_by_current_user user.favourite_field_ids)
+        .not_to include(field_3)
+      end
+    end
+
+    context "when falsy or no argument provied" do
+      it "returns empty" do
+        expect(Field.favourite_by_current_user).to be_empty
+        expect(Field.favourite_by_current_user nil).to be_empty
+      end
+    end
+  end
+
+  describe "#average_rating" do
+    let!(:field_1){create :field}
```

Comment:
phan biet let vs let!

Reply:
- let: sẽ không gọi và gán giá trị cho biến ngay mà đợi đến khi method it đầu tiên được gọi thì mới chạy
- let!: gọi và gán giá trị cho biến ngay lập tức